File: clientcpp/Services/UDMServer.cpp
//---------------------------------------------------------------------------


#pragma hdrstop

#include "UDMServer.h"
#include "UUtilities.h"
#include "ExceptionHandler.h"
#include "../UHelper.hpp"

#include <FMX.Dialogs.hpp>
//---------------------------------------------------------------------------
#pragma package(smart_init)
#pragma classgroup "FMX.Controls.TControl"
#pragma resource "*.dfm"
TdmData *dmData;
//---------------------------------------------------------------------------
__fastcall TdmData::TdmData(TComponent* Owner)
	: TDataModule(Owner)
	, _user(new User())
{
}
//---------------------------------------------------------------------------

UPtrJSONObject TdmData::ExecREST(String method, const UPtrJSONObject& body)
{
    std::lock_guard<std::mutex> lock(_m_execrest);
	std::unique_ptr<TRESTClient> rClient(new TRESTClient(this));
	std::unique_ptr<TRESTRequest> rRequest(new TRESTRequest(this));
	std::unique_ptr<TRESTResponse> rResponse(new TRESTResponse(this));
	rClient->BaseURL = "http://localhost:3000/api/sm/";
	rRequest->Client = rClient.get();
	rRequest->Response = rResponse.get();
	rRequest->Method = TRESTRequestMethod::rmPOST;
	rRequest->Resource = method;
	rRequest->SynchronizedEvents = false;
	rRequest->Timeout = 10000;
	rRequest->Body->Add(body.get());
	int codError = 0;
	try
	{
		rRequest->Execute();
		if (!rResponse->JSONValue->TryGetValue<int>("codError", codError))
		{
			throw new Exception("Invalid server response");
		}

		TJSONObject *resobj = dynamic_cast<TJSONObject*>(rResponse->JSONValue->Clone());
		UPtrJSONObject resptr(resobj);
		return std::move(resptr);
	}
	catch (Exception& e)
	{
		UPtrJSONObject mr(new TJSONObject());
		mr->AddPair("codError", IntToStr(CONNECTION_ERROR));
		return std::move(mr);
	}
}
//---------------------------------------------------------------------------

ExceptionHandler TdmData::Login(refStr nickname, refStr password)
{
	UPtrJSONObject body(new TJSONObject());
	UPtrJSONObject res;
	String pass = MakeHash512(password);
	body->AddPair("nickname", nickname.LowerCase());
	body->AddPair("password", pass);
	res.reset(ExecREST("login", body).release());
	int codError = StrToInt(res->GetValue("codError")->Value());

	try
	{
		if (codError)
			throw ExceptionHandler();

		_user->Id = StrToInt(res->GetValue("id")->Value());
		_user->Nickname = nickname;
		_user->Password = pass;
		_user->Token = res->GetValue("token")->Value();

		GetUserPhoto(_User()->Id);

		return ExceptionHandler(true);
	}
	catch (ExceptionHandler& e)
	{
		return ExceptionHandler(false, codError, "TdmData::Login");
    }
}
//---------------------------------------------------------------------------

ExceptionHandler TdmData::SignIn(refStr nickname, refStr lastname,
						refStr firstname, refStr email, refStr password, TBitmap *bm)
{
    UPtrJSONObject body(new TJSONObject());
	UPtrJSONObject res;
	String pass = MakeHash512(password);
	body->AddPair("nickname", nickname.LowerCase());
	body->AddPair("lastname", lastname.LowerCase());
	body->AddPair("firstname", firstname.LowerCase());
	body->AddPair("email", email);
	body->AddPair("password", pass);
	std::unique_ptr<TBytesStream> b(new TBytesStream(TBytes()));
	bm->SaveToStream(b.get());
	body->AddPair("image", TEncryp::B64Encode(b.get()));
	res.reset(ExecREST("signin", body).release());
	int codError = StrToInt(res->GetValue("codError")->Value());

	try
	{
		if (codError)
			throw ExceptionHandler();

		return ExceptionHandler(true);
	}
	catch (ExceptionHandler& e)
	{
		return ExceptionHandler(false, codError, "TdmData::SignIn");
	}
}
//---------------------------------------------------------------------------

ExceptionHandler TdmData::GetUserPhoto(int id)
{
	UPtrJSONObject body(new TJSONObject());
	UPtrJSONObject res;
	body->AddPair("id", _User()->Id);
	body->AddPair("token", _User()->Token);
	body->AddPair("id_usr", id);
	res.reset(ExecREST("getuserphoto", body).release());
    int codError = StrToInt(res->GetValue("codError")->Value());

    try
	{
		if (codError)
			throw ExceptionHandler();

		std::unique_ptr<TBytesStream> b(TEncryp::B64Decode(res->GetValue("image")->Value()));
        b->Position = 0;
		std::unique_ptr<TBitmap> bmp(new TBitmap());
		bmp->LoadFromStream(b.get());
		_user->Image = bmp.release();
		return ExceptionHandler(true);
	}
	catch (ExceptionHandler& e)
	{
		return ExceptionHandler(false, codError, "TdmData::GetUserPhoto");
	}
}
//---------------------------------------------------------------------------

void __fastcall TdmData::DataModuleDestroy(TObject *Sender)
{
	if (_user) delete _user;
	_user = nullptr;
}
//---------------------------------------------------------------------------

User* const TdmData::_User() const
{
    return _user;
}
//---------------------------------------------------------------------------

File: clientpas/UDMServer.pas
unit UDMServer;

interface

uses
  System.SysUtils, System.Classes, System.JSON,
  Data.Bind.Components, Data.Bind.ObjectScope,
  REST.Client, REST.Types, IPPeerClient, UUtilities,
  FMX.Graphics, UHelper, UExceptionHandler, System.SyncObjs,
  // Models
  UModels;

type
  TdmData = class(TDataModule)
    procedure DataModuleCreate(Sender: TObject);
    procedure DataModuleDestroy(Sender: TObject);
  private
    var
      FUser: TUser;

    procedure ExecRest(method: string; body: TJSONObject; out res: TJSONObject);
  public
    // method - Session
    procedure LogIn(nickname, password: string; var exce: TExceptionHandler);
    procedure SignIn(nickname, firstname, lastname, email, password: string;
                    bm: TBitmap; var exce: TExceptionHandler);

    destructor Destroy; virtual;

    // getter
    procedure GetUserPhoto(id: Integer; out stream: TBytesStream);

    // property
    property User: TUser read FUser write FUser;
  end;

var
  dmData: TdmData;

implementation

{%CLASSGROUP 'FMX.Controls.TControl'}

{$R *.dfm}

{ TdmData }

procedure TdmData.DataModuleCreate(Sender: TObject);
begin
  FUser := TUser.Create;
end;

procedure TdmData.DataModuleDestroy(Sender: TObject);
begin
  FUser.DisposeOf;
end;

destructor TdmData.Destroy;
begin
  //
end;

procedure TdmData.ExecRest(method: string; body: TJSONObject; out res: TJSONObject);
var
  client: TRESTClient;
  request: TRESTRequest;
  response: TRESTResponse;
  codError: Integer;
  obj: TJSONObject;
begin
  client := TRESTClient.Create(Self);
  request := TRESTRequest.Create(Self);
  response := TRESTResponse.Create(Self);
  client.BaseURL := 'http://localhost:3000/api/sm/';
  request.Client := client;
  request.Response := response;
  request.Method := rmPOST;
  request.Resource := method;
  request.SynchronizedEvents := False;
  request.Timeout := 10000;
  request.Body.Add(body);
  try
    request.Execute;
    if response.JSONValue.TryGetValue('codError', codError) = False then
      raise Exception.Create('Invalid server response');

    obj := TJSONObject(response.JSONValue.Clone);
  except
    on e: Exception do
    begin
      obj := TJSONObject.Create;
      obj.AddPair('codError', '-999');
    end;
  end;

  res := obj;

  Client.Free;
  Request.Free;
  Response.Free;
end;

procedure TdmData.GetUserPhoto(id: Integer; out stream: TBytesStream);
var
  body: TJSONObject;
  res: TJSONObject;
  codError: Integer;
begin
  body := TJSONObject.Create;
  body.AddPair('id', IntToStr(User.Id));
  body.AddPair('token', User.Token);
  body.AddPair('id_usr', IntToStr(id));
  ExecRest('getuserphoto', body, res);
  codError := StrToInt(res.GetValue('codError').Value);

  try
    if codError <> 0 then
      raise Exception.Create('GetUserPhoto');

    B64Decode(res.GetValue('image').Value, stream);
  finally
    body.Free;
    res.Free;
  end;
end;

procedure TdmData.LogIn(nickname, password: string; var exce: TExceptionHandler);
var
  body: TJSONObject;
  res: TJSONObject;
  pass_hash: string;
  codError: Integer;
begin
  body := TJSONObject.Create;
  body.AddPair('nickname', nickname);
  pass_hash := MakeHash512(password);
  body.AddPair('password', pass_hash);
  ExecRest('login', body, res);
  codError := StrToInt(res.GetValue('codError').Value);
  exce.CodError := codError;
  try
    if codError <> 0 then
      raise Exception.Create('LogIn');

    User.Id := StrToInt(res.GetValue('id').Value);
    User.Nickname := nickname;
    User.Password := pass_hash;
    User.Token := res.GetValue('token').Value;

    exce.Response := True;
  except
    on e: Exception do
    begin
      exce.Response := False;
    end;
  end;

  body.Free;
  res.Free;
end;

procedure TdmData.SignIn(nickname, firstname, lastname, email, password: string;
                    bm: TBitmap; var exce: TExceptionHandler);
var
  body: TJSONObject;
  res: TJSONObject;
  pass_hash: string;
  codError: Integer;
  b: TBytesStream;
  bytes: TBytes;
begin
  body := TJSONObject.Create;
  body.AddPair('nickname', nickname);
  body.AddPair('lastname', lastname);
  body.AddPair('firstname', firstname);
  body.AddPair('email', email);
  pass_hash := MakeHash512(password);
  body.AddPair('password', pass_hash);
  b := TBytesStream.Create(bytes);
  bm.SaveToStream(b);
  body.AddPair('image', B64Encode(b));
  ExecRest('signin', body, res);
  codError := StrToInt(res.GetValue('codError').Value);
  exce.CodError := codError;

  try
    if codError <> 0 then
      raise Exception.Create('SignIn');

    exce.Response := True;
  except
    on e: Exception do
    begin
      exce.Response := False;
    end;
  end;

  b.Free;
  body.Free;
  res.Free;
end;

end.
